src/kubernetes_cluster/proof/controller_runtime_safety.rs

```diff
@@ -513,4 +514,90 @@ pub proof fn lemma_always_each_resp_matches_at_most_one_pending_req<K: ResourceV
     init_invariant::<State<K, T>>(sm_spec::<K, T, ReconcilerType>(), init::<K, T, ReconcilerType>(), stronger_next, invariant);
 }
 
+// This lemma ensures that if a controller is at some reconcile state for a cr, there must be the pending request of the
+// reconcile state in flight or a correponding response in flight.
+// Obviously, this requires that when controller enters the 'state' in reconcile_core, there must be a request generated;
+// otherwise, the pending request may not be there.
+// The proof is very straightforward:
+//   - Right after the controller enters 'state', the pending request is added to in_flight.
+//   - If the pending request is processed by kubernetes api, there will be a response in flight.
+//   - If the response is processed by the controller, the controller will either enter a new state or a new pending request in flight.
```

Comment:
> either enter a new state or a new pending request in flight

Or?

Reply:
Yes. By this I mean the controller will not necessarily enter a new state; if it enters a state that is same as 'state', then a new pending request will be in flight, so in this case the invariant will not be violated.

Reply:
I see what you mean. Maybe it is better to say "If the response is processed by the controller, the controller will either the next state (which could still be 'state' but the invariant still holds then)."